src/Elastic.Apm.SqlClient/SqlEventListener.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+
+namespace Elastic.Apm.SqlClient
+{
+	internal class SqlEventListener : EventListener
+	{
+		private readonly ApmAgent _apmAgent;
+		private readonly IApmLogger _logger;
+
+		private readonly ConcurrentDictionary<int, (Span Span, long Start)> _spans = new ConcurrentDictionary<int, (Span, long)>();
+
+		public SqlEventListener(IApmAgent apmAgent)
+		{
+			_apmAgent = (ApmAgent)apmAgent;
+			_logger = _apmAgent.Logger.Scoped(nameof(SqlEventListener));
+		}
+
+		protected override void OnEventSourceCreated(EventSource eventSource)
+		{
+			if (eventSource != null && eventSource.Name == "Microsoft-AdoNet-SystemData")
```

Comment:
I saw some strange behavior around here. The `OnEventSourceCreated` method got triggered twice for me resulting with the following `Paylaod` in the `OnEventWritten` method:

```
[0] = "ERROR: Exception in Command Processing for EventSource Microsoft-AdoNet-SystemData: An instance of EventSource with Guid 6a4dfe53-eb50-5332-8473-7b7e10a94fd1 already exists."
```

So, in the `OnEventWritten` we never did anything since the payload was not something we expect there. 

My hack to make that work was to disable the eventsource and enable it again - I don't know why this happens and I don't really like this code - I'm not sure what should be the solution here. 

```
if (eventSource != null && eventSource.Name == "Microsoft-AdoNet-SystemData")
{
	if (_eventSource != null)
	{
		DisableEvents(_eventSource);
		_eventSource.Dispose();
	}

	EnableEvents(eventSource, EventLevel.Informational, (EventKeywords)1);
	_eventSource = eventSource;
}
```
(`_eventSource` is a field on `SqlEventListener`) 

One suspect is that both `System.Data` and `Microsoft.Data` triggers this.

Reply:
Thanks for your findings. I'll look deeper into it.

Reply:
@gregkalapos, your assumption is right. There are 2 event sources with the same GUIDs (GUID is generated based on event source name), one from `System.Data.SqlClient` and another one from `Microsoft.Data.SqlClient`.
This is the exact place where the exception is thrown: https://github.com/microsoft/referencesource/blob/08b84d13e81cfdbd769a557b368539aac6a9cb30/mscorlib/system/diagnostics/eventing/eventsource.cs#L3029-L3033
With this place, it's understandable why disposing of previous event source help to solve the issue, however, it's quite unstable, sometimes it doesn't help.
I've created [an issue in `Microsoft.Data.SqlClient` repo](https://github.com/dotnet/SqlClient/issues/436). Will look, maybe they have a proper solution.

For now, I suggest keeping it as is with small hacking for our tests because I don't think that a lot of people use both libraries in one application (maybe I'm wrong). WDYT?

Reply:
>  I suggest keeping it as is with small hacking for our tests because I don't think that a lot of people use both libraries in one application

Yes, I agree that the point of this discussion is not a very "real world" scenario - I don't think either people would have both in 1 app. 

On the other hand - what's the small hack for the tests? Did you already push something for this? 🤔  Currently the tests on full framework fail to me due to this problem.

Reply:
Yes, I haven't committed the hack yet. An idea is pretty simple. As `SqlEventSource` initializes statically the only one possible approach is to have three solutions. One with a base abstract class which will have test definitions, another one with referencing `System.Data.SqlClient` and the last one with `Microsoft.Data.SqlClient`.
I know, it's not a very good solution, but we will have it until [version `2.0.0` released with a fix for event source name](https://github.com/dotnet/SqlClient/issues/436#issuecomment-588434098)

Reply:
@gregkalapos, I tried to implement solution with 3 projects but it also doesn't work. The issue is the `SqlEventSource` for `System.Data.SqlClient` is situated in the `System.Data` assembly. It means that `SqlEventSource` is always instantiated even if we don't reference `System.Data.SqlClient`.
So, my final suggestion (I believe) is to skip instrumentation for `Microsoft.Data.SqlClient` in .Net Framework. WDYT?

Reply:
> So, my final suggestion (I believe) is to skip instrumentation for Microsoft.Data.SqlClient in .Net Framework. WDYT?

Ok, fine by me. 

Overall, I think having 3 projects wouldn't even be a hack, more like a legitimate solution - it's unfortunate that it does not work.

Anyways, let's skip it for Full Framework and see if someone needs it in the future.

Thanks for opening [the issue](https://github.com/dotnet/SqlClient/issues/436) in the SqClient repo.

Nice to have the response there, but having yet another name in the future (as the comment says) makes this thing super complicated - anyways this is outside our control, so we use whatever they offer...

Reply:
> but having yet another name in the future (as the comment says) makes this thing super complicated